keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,262 @@
+---
+title: Even Pods Spreading
+authors:
+  - "@Huang-Wei"
+owning-sig: sig-scheduling
+reviewers:
+  - "@wojtek-t"
+  - "@bsalamat"
+  - "@k82cn"
+approvers:
+  - "@bsalamat"
+  - "@k82cn"
+creation-date: 2019-02-21
+last-updated: 2019-02-27
+status: provisional
+---
+
+# Even Pods Spreading
+
+## Table of Contents
+
+* [Terms](#terms)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+    * [Story 1 - NodeAffinity](#story-1---nodeaffinity)
+    * [Story 2 - PodAffinity](#story-2---podaffinity)
+    * [Story 3 - PodAntiAffinity](#story-3---podantiaffinity)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Design Details](#design-details)
+  * [Algorithm](#algorithm)
+    * [NodeAffinity](#nodeaffinity)
+    * [PodAffinity](#podaffinity)
+    * [PodAntiAffinity](#podantiaffinity)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Terms
+
+- **Topology:** describe a series of worker nodes which belongs to the same
+  region/zone/rack/hostname/etc. In terms of Kubernetes, they're defined and
+  grouped by node labels.
+- **Affinity**: if not specified particularly, "Affinity" refers to
+  `NodeAffinity`, `PodAffinity` and `PodAntiAffinity`.
+- **CA**: Cluster Autoscaler. [CA](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is a tool that automatically adjusts the size of the Kubernetes cluster upon specific conditions.
+
+## Summary
+
+`EvenPodsSpreading` feature applies on NodeAffinity/PodAffinity/PodAntiAffinity
```

Comment:
Why are we introducing this feature as a "sub-feature" of other existing affinities? I think this should be its own independent feature.
A pod spec may specify even spreading. Even spreading can be "required" or "preferred". Topology key works exactly like topology key in our inter-pod affinity rules. We can have "MaxSkew" exactly like you described. We also need a namespace and label selector very similar to pod affinity. I believe this should cover all use-cases well. Pod spec may still have other affinity rules if further placement criteria need to be specified.

Reply:
@bsalamat thanks for the reviewing!

Here are my thoughts on putting this feature as **a)** a "sub-feature" onto pod.spec.affinity versus **b)** an independent feature onto pod.spec:

1. **Duplications.** In user's perspective, if "even spreading" feature applies to pod.spec, it would concern the users since this looks exactly the same as pod.spec.podAffinity spec. People would question why a "duplicated" API design is introduced. And the duplication would be more severe if both "required" and "preferred" are injected into pod.spec.
Additionally, internally we have to maintain a similar structure in PredicateMetadata (where affinity is pre-handled) to cover the "even spreading", that would lead to duplicate code like once a pod event is notified, we need to manipulate two (or more) structures multiple times, and the processing logic in predicates/priority phase may be duplicated as well. IMO this looks not neat and would increase memory footprint.

1. **Usecase combinations become a matrix.** If go with an independent feature onto pod spec, all combinations of possible usecases would be (even spreading enabled or not) * (required or preferred) * (NodeAffinity/PodAffinity/PodAntiAffinity) * (required or preferred on Affinity). It's way too complicated. (It's even a little hard for me to figure out the exact meaning of some combination...)
And after taking a first glance at the matrix, "Story 3 - PodAntiAffinity" seems not to be covered unless we put particular "hacking" code. I'm a little  concerned on making code this way to support the full matrix.